services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusUpdateRequestValidator.kt

```diff
@@ -60,18 +60,72 @@ fun CustomerRequest.validateForErrors(path: String): ValidationResult =
     validate { notJustWhitespace(name, "$path.name") }
         .and { notJustWhitespace(email, "$path.email") }
         .and { notJustWhitespace(phone, "$path.phone") }
-        .andWhen(
-            registryKey != null &&
-                (type == CustomerType.COMPANY || type == CustomerType.ASSOCIATION)
-        ) {
+        .and { validateRegistryKey(path) }
+
+fun CustomerRequest.validateRegistryKey(path: String): ValidationResult =
+    when (type) {
+        CustomerType.COMPANY ->
+            validateRegistryKeyForCompanies(this.registryKey, this.registryKeyHidden, path)
+        CustomerType.ASSOCIATION ->
+            validateRegistryKeyForCompanies(this.registryKey, this.registryKeyHidden, path)
+        CustomerType.PERSON ->
+            validateRegistryKeyForPerson(this.registryKey, this.registryKeyHidden, path)
+        CustomerType.OTHER ->
+            validateRegistryKeyForOther(this.registryKey, this.registryKeyHidden, path)
+    }
+
+private fun validateRegistryKeyForCompanies(
+    registryKey: String?,
+    registryKeyHidden: Boolean,
+    path: String,
+) =
+    validateFalse(registryKeyHidden, "$path.registryKeyHidden").and {
+        whenNotNull(registryKey) {
             validateTrue(registryKey.isValidBusinessId(), "$path.registryKey")
         }
+    }
+
+private fun validateRegistryKeyForOther(
+    registryKey: String?,
+    registryKeyHidden: Boolean,
+    path: String,
+) =
+    validate()
+        .andWhen(registryKeyHidden) { validateNull(registryKey, "$path.registryKey") }
+        .andWhen(!registryKeyHidden) {
+            whenNotNull(registryKey) { notBlank(it, "$path.registryKey") }
+        }
+
+private fun validateRegistryKeyForPerson(
+    registryKey: String?,
+    registryKeyHidden: Boolean,
+    path: String,
+) =
+    validateRegistryKeyForOther(registryKey, registryKeyHidden, path).whenNotNull(registryKey) {
```

Comment:
I see some possibility of errors here: if there is a change in logic in `validateRegistryKeyForOther` this might break. Maybe just copy-paste the logic here or have some common function used in both validations?

Reply:
Thank you for the suggestion. This turned out better after I inlined the `validateRegistryKeyForOther` method here. The henkilötunnus validation was done even when not necessary and running both `notBlank` and `isValidHenkilotunnus` was redundant. These also contributed to the double error paths you reported in the other PRs.

Solving this didn't get rid of all the double errors, so I changed the validation to return only unique error paths. Some changes were needed in other parts of the code because of this.